### PR TITLE
Expose Cloud Build Options and hide tabs and features

### PR DESCRIPTION
**Pull request by haslinghuis**
- [x] Keeps default behavior when BUILD_KEY or internet connection is not available
- [x] Retrieve cloud build options using BUILD_KEY in auto-detect and serial backend
- [x] Hides features which are not in cloud build (configuration tab and telemetry for receiver tab)
- [x] Shows tabs which are in cloud build and hide if not.

Please test with firmware https://github.com/betaflight/betaflight/pull/12388 as this would address:
- [x] features can be enabled but we still have to save to eeprom to activate - this should be done in firmware with cloud build options but not sure how we macro that up with DEFAULT_FEATURES

:thinking: Future improvements:

- [ ] moving feature switch to respective tab
- [ ] having port configuration on each tab with label for function, if there is already a function a warning will indicate saving will disable the previous function.
- [ ] receiver configuration
- [ ] motor protocol configuration

As example a very simple build for a LOS build:

image

we see:

image

As for a more sophisticated build like:

image

we see:

image

Originally we see features which are not available in firmware:

image

### CODE REVIEW

File: src/js/fc.js
Comment (blckmn) [inline src/js/fc.js:11]: capital first letter?

File: src/js/serial_backend.js
Comment (blckmn) [inline src/js/serial_backend.js:366]: this should probably be called as part of the autoDetect method?

Comment (blckmn) [reply src/js/serial_backend.js:366]: https://github.com/betaflight/betaflight-configurator/blob/309bd431289df96c75092c163b15e378e441cae8/src/js/tabs/firmware_flasher.js#L579

**Review by blckmn — CHANGES_REQUESTED** [review]

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

File: src/js/serial_backend.js
Comment (haslinghuis) [reply src/js/serial_backend.js]: Added to autoDetect

File: src/js/fc.js
Comment (haslinghuis) [reply src/js/fc.js]: Fixed.

File: src/js/BuildApi.js
Comment (chmelevskij) [inline src/js/BuildApi.js:161]: Can we promisify this instead then just use `try/catch` or `then/catch`?

File: locales/en/messages.json
Comment (McGiverGim) [inline locales/en/messages.json:1069]: If for the user this is more a "Sonar", maybe is better to rename it in the firmware and not here?

File: src/js/Features.js
Comment (McGiverGim) [inline src/js/Features.js:44]: I think we are having too much hardcoded elements. This will require to maintain in coherence firmware/configurator. I understand to do this for features, that is something that we are trying to remove so it will not grow, but I don't know if we are planning to do the same for other UI elements.

File: src/js/BuildApi.js
Comment (McGiverGim) [reply src/js/BuildApi.js:161]: We are getting the features from the cloud build information? This is not the way. What if the build is very old, we will store ALL the builds in the time? What if the build is a local custom build? If we want to get this information, I think we need to create MSP commands in the firmware, that can tell us what features/other UI elements are enabled or not.

Comment (haslinghuis) [reply src/js/BuildApi.js:161]: @chmelevskij was just following same pattern in the api.
@McGiverGim there is a msp call to retrieve the buildkey. With the buildkey we request data from the cloud api filling some MSP variables. We might be able to store the filter in the cloud api. We only can use this from semver 1.45.
Local builds will have no buildkey (need to check).
I'd like to lock some configuration fields with cloud parameters as to hide items that are not included in the cloud build but the user still was able to select. Thoughts?

Comment (McGiverGim) [reply src/js/BuildApi.js:161]: As I say, in my opinion, the configurator must work with customs builds too, but the change to cloud build is a very big step and the Configurator had left outdated showing features that does not exist.
Maybe with the new MSP call with CLI we can ask for features to the CLI, if it shows an error then does not exist.

Comment (blckmn) [reply src/js/BuildApi.js:161]: I think there's two different issues here.

The first is using the build key, the user can default their build options to that contained for that key. Key not being user specific but build specific. Without key (custom or key not found) the build options revert to default.

The second is what actual features in configurator are enabled. Eg don't show GPS tab if GPS not built in. That should come from the firmware. Using the cli suggestion is good.

File: locales/en/messages.json
Comment (haslinghuis) [reply locales/en/messages.json:1069]: Good call, will change RANGEFINDER to SONAR in firmware but RANGEFINDER is just one possible SONAR device?

File: src/js/Features.js
Comment (haslinghuis) [reply src/js/Features.js:44]: Yes this is hack, becomes the 'other' group contains also non-features. While features are very stable over time this PR is just enabling support for cloud parameters AND applying this to features as a demonstration.

**Comment by haslinghuis** [comment]
IGNORE THIS COMMENT :)
```
        self._features = features.map(feature => {
            return {
                ...feature,
                available: filter.some(opt => opt.includes(feature.name)) ? FC.CONFIG.buildOptions.some(opt => opt.includes(feature.name)) : true,
            };
        }).sort((a, b) => a.name.localeCompare(b.name, window.navigator.language, { ignorePunctuation: true }));

```

**Comment by haslinghuis** [comment]
This works:
```
        for (const f of features) {
            if (FC.CONFIG.buildOptions.some(opt => opt.includes(f.dependsOn)) || f.dependsOn === undefined) {
                console.log('Adding:', f.name);
                self._features.push(f);
            }
        }
```

But this does not:
```
        self._features = features.filter(feature => FC.CONFIG.buildOptions.some(opt => opt.includes(feature.dependsOn) || feature.dependsOn === undefined))
```

**Comment by haslinghuis** [comment]
Again: this happened so many times last year :(
```
Error building NW apps: Error: connect ETIMEDOUT 45.79.75.204:443
```

File: src/js/Features.js
Comment (haslinghuis) [reply src/js/Features.js]: Have removed the hack and added a property to the feature instead.

**Review by blckmn — APPROVED** [review]

File: src/js/BuildApi.js
Comment (haslinghuis) [reply src/js/BuildApi.js]: @blckmn sorted now:

Buildkey availability determines behavior.
The tab issue is also resolved now.

**Comment by nerdCopter** [comment]
LED tab not showing even with LED toggle enabled.

**Comment by nerdCopter** [comment]
something not right with some builds -- tabs dont show even with toggles enabled.
image

**Comment by haslinghuis** [comment]
@nerdCopter some glitch. It's working now with improvements. Need to reflash firmware.

**Comment by nerdCopter** [comment]
* `9f0cbb2f` tested good :)

**Comment by nerdCopter** [comment]
* Tested good with new GPS logic.

**Comment by haslinghuis** [comment]
@McGiverGim 

> We are getting the features from the cloud build information? This is not the way. What if the build is very old, we will store ALL the builds in the time? What if the build is a local custom build? If we want to get this information, I think we need to create MSP commands in the firmware, that can tell us what features/other UI elements are enabled or not.

Agree this should be explored. As with the buildKey stored in firmware we can retrieve build options from cloud build. Having them set in firmware removes the dependency of needing an internet connection. For now it there is no build key or internet connection we just show the full set.

**Comment by McGiverGim** [comment]
Another con is that the user sees unexpected behavior, depending on the board, if it has internet or not, etc.

**Comment by nerdCopter** [comment]
* Seems to be populating cloud-build options properly with exception of the `CRSF,GHST,SBUS` multi-option which populates as SBUS.
* Maybe the `Radio Protocol` field can be reverted to the same style as `Other Options`. @blckmn

**Comment by haslinghuis** [comment]
The `CRSF,GHST,SBUS` should be removed as everybody should know some basics.

**Comment by nerdCopter** [comment]
does this need some `&&` or `||` logic for older firmware? or is backward compatibility not on the agenda here?
Untitled2

**Comment by haslinghuis** [comment]
Thanks, needs updating semver to 1.46 instead.

**Comment by haslinghuis** [comment]
@nerdCopter please check again.

🤔 We might decouple features toggling tabs (like GPS) to being enabled by default (if GPS option is included in build)

**Comment by nerdCopter** [comment]
i'm seeing same behavior as last screenshot for 4.4.0. However, i'm not sure what to expect.
I cannot seem to test `master` at this time as it's broken for my unsoldered TMOTORF7 (`Failed to open serial port` after flash)

**Comment by haslinghuis** [comment]
@nerdCopter https://github.com/betaflight/betaflight/pull/12308 was the last working for boards having same issue (all leds are solid indicating hard fault)

**Comment by nerdCopter** [comment]
> @nerdCopter [betaflight/betaflight#12308](https://github.com/betaflight/betaflight/pull/12308) was the last working for boards having same issue (all leds are solid indicating hard fault)

unfortunately #12308 not working for TMOTORF7 :(   i'll pan through some older SHA's

**Comment by haslinghuis** [comment]
#12388 needs a redesign now we are removing custom defaults.

**Comment by nerdCopter** [comment]
* Does not seem to be working anymore. firmware `master`. flashed with `crsf`,`flash`,`osd sd`,`vtx`. populated defaults upon clicking `auto-detect`.
* gps and led still available
* no console errors. 
image
imageimage

**Comment by nerdCopter** [comment]
sorry @haslinghuis , i'm deleting last comment as it checks for 1.46 now. unsure how to further test without a live 1.46.

**Comment by haslinghuis** [comment]
You could test with commit `#12425`

**Comment by nerdCopter** [comment]
`#12425` is still msp 1.45 unfortunately. suppose i can mod locally for the purposes of testing.

**Comment by haslinghuis** [comment]
It works with `#12323` (after applying defaults)

**Comment by nerdCopter** [comment]
* tested with #12323 successfully -- however, apply custom defaults breaks my tmotorf7 serial
* tabs of existing features remain visible irregardless of toggle position. is this the desired logic?
* servos tab visible even without feature
image

**Comment by haslinghuis** [comment]
@nerdCopter fixed the tab issue (inverting check and make available for 1.45)

**Comment by nerdCopter** [comment]
*needs hide/show for servos toggle, then i think it's good to go.
* caveat is the `CRSF,GHST,SBUS` option which i suppose needs to be a separate PR.  want to re-iterate i'd prefer same look&feel as custom defines (bubbles)

**Review by nerdCopter — APPROVED** [review]
💪🚀

**Pull request merged**
